### ISSUE TITLE
Tpetra and Xpetra:  request for typedef or clarification of offset_type

### ISSUE THREAD

**Issue by kddevin** [trilinos/Trilinos#674]
Tpetra uses offset_type for the offsets into the column pointers array.  From #153, it is difficult to tell whether this type is always size_t, or is sometimes int, or is sometimes defined by the node.  I am further confused because Tpetra::CrsMatrix::getAllValues uses size_t for the rowPointers; does it make a copy when offset_type is not size_t?  Does getAllValues() make copies of all three arrays?

Should we assume that getAllValues will always return size_t rowPointers?  If not, we'll need Tpetra and Xpetra to add a public typedef allowing users to access the offset_type, similar to local_ordinal_type or global_ordinal_type.  

This request supports #665.  In Zoltan2, we had assumed offsets would be local_ordinal_type.  In order to use views of the offsets (instead of making copies in Zoltan2), we need to change our offsets to Tpetra's offset type.  But before we propagate this change through Zoltan2, we'd like to understand what offset_type really is, and whether we should use size_t everywhere or some other offset_type.  This is a change we'd like to propagate only once, so if there are plans for changes in the future, please let us know.  

Thanks.
@trilinos/tpetra @trilinos/xpetra @trilinos/zoltan2

**Comment by mhoemmen** [comment]
@kddevin 
> ... it is difficult to tell whether this type is always size_t, or is sometimes int, or is sometimes defined by the node.

It is a function of the device type.  With Kokkos::Cuda, it is int.  With non-CUDA device types, it is size_t.  We brought up at the recent Kokkos API Review making size_type always a signed integer (size_t is generally unsigned).

> I am further confused because Tpetra::CrsMatrix::getAllValues uses size_t for the rowPointers; does it make a copy when offset_type is not size_t?

Yes.  You should consider both getAllValues and setAllValues as methods that exist only for backwards compatibility.

> Does getAllValues() make copies of all three arrays?

No.  The returned Teuchos::ArrayRCP objects just wrap the matrix's Kokkos::View, and have special destructors that do nothing but hang on to the Kokkos::View instance (thus letting Kokkos do reference counting for its View objects).  However, setAllValues _must_ (unfortunately) make deep copies of all three arrays, since Kokkos::View has no way to take ownership of input memory from a Teuchos::ArrayRCP.

> If not, we'll need Tpetra and Xpetra to add a public typedef allowing users to access the offset_type ...

Tpetra has this already: `typename Tpetra::CrsMatrix<...>::local_matrix_type::size_type`.

**Comment by mhoemmen** [comment]
> Tpetra has this already: typename Tpetra::CrsMatrix<...>::local_matrix_type::size_type.

That did sound a bit rude, didn't it? :-)  It looks like you would like a typedef `offset_type` in CrsMatrix.  This sounds like a reasonable request.

**Comment by srajama1** [comment]
Is this complication really necessary ? In addition to making it a signed integer, can we just get rid of it and use one of lno_t and gno_t (pick one) ? When do we really need to use >2B nnz within an MPI rank ? I missed this portion of the review, so excuse me if this has been already discussed.

**Comment by mhoemmen** [comment]
> Is this complication really necessary ? 

  1. offset_type != local_ordinal_type: Mike Heroux has asked for it to be that way for years.  Christian also wants it.  Makes sense since we'll have nodes with ~ 1 TB memory.

  2. offset_type != global_ordinal_type: KokkosKernels is local.  It has no idea about "global" ordinals.  Thus, KokkosKernels must pick something, at least by default.  Furthermore, setting offset_type == global_ordinal_type reduces to the above argument (1) if local_ordinal_type == global_ordinal_type == int.

  3. Overall effort to support offset_type != local_ordinal_type is generally much lower than the time spent complaining about it ;-) 

> When do we really need to use >2B nnz within an MPI rank ?

See above.

**Comment by mndevec** [comment]
> 
>    
Does getAllValues() make copies of all three arrays?
>
No. The returned Teuchos::ArrayRCP objects just wrap the matrix's Kokkos::View, and have special destructors that do nothing but hang on to the Kokkos::View instance (thus letting Kokkos do reference counting for its View objects). 
> 

Mark, 
Just to be clear, it does not make any deep copies if offset_type is size_t. But I guess it has to do a deep copy if offset_type is int for Cuda, right?

**Comment by mhoemmen** [comment]
> Just to be clear, it does not make any deep copies if offset_type is size_t. But I guess it has to do a deep copy if offset_type is int for Cuda, right?

Yes.  It's just a deep copy of the row offsets, not of anything else.

**Comment by maherou** [comment]
@srajama1 Just to follow up on @mhoemmen 's comment: >2GB is a near-term reality for some of our users.  We really do need it.

**Comment by maherou** [comment]
@srajama1 @mhoemmen : Sorry, 2B nz.

**Comment by kddevin** [comment]
Hi, @mhoemmen .
1.  If getAllValues exists only for backward compatibility,  should we be using something else instead to get the three pointers?   Is there a mechanism that would always avoid the copy?  We'd be happy to get the pointer to the offset_type array, rather than have it converted to size_t.  Something in Xpetra would be ideal, but we could start with Tpetra.
2.  A typedef would be great, as we could use it to do the propagation only once.  Again, in Xpetra would be best, but we could start with Tpetra.
Yesterday, I forgot to cite @krcb @MicheldeMessieres.

**Comment by srajama1** [comment]
Alright, I will leave complaining about it. This might be easy to support as Mark says. However, I hope this is not one off. Extensions like these require long term maintenance, real costs in propagating through the stack, compile times and above all usability. Will it be that hard to say if you have more than 2B nnz run two MPI ranks ? It might scale well with that large of a problem anyway.

**Comment by nmhamster** [comment]
@srajama1 need to remember there are systems on the horizon which will have quite large memory versus today (certainly with respect to number of MPI ranks per node) and in the future this could be considerably larger. If it does take a long time to propagate through the stack I would advocate for at least having the discussions on how this could be supported now. We can talk offline about some of the details if you would like.

**Comment by egboman** [comment]
@nmhamster , For such large systems with large memory then one could simply let local ordinals, global ordinals and offset_type all be int64_t (long long)? This works now, I believe. This whole discussion is on the narrower use case where the number of nonzeros is > 2B but the number of rows/columns is not.

**Comment by nmhamster** [comment]
@egboman agreed that for the largest this makes sense but I would conjecture that there will be systems in the middle. I guess the question in whether running entirely with 64-bit significantly affects performance enough to be concerned (when 32-bit might be possible for rows/columns but not NNZ). Given that it affects bandwidth the answer is likely yes but maybe not enough that we would care to develop/maintain special versions of code to handle this. I was merely pointing out that we are very likely to have some systems in this bracket (MPI core to memory ratios).

**Comment by maherou** [comment]
@egboman If the transition from 32-bit ints to 64-bit ints in the global ID space is any indication, we had about 5 - 7 years where we had to report the global number of nonzero as a "long long" before we had to transition to 64-bit global IDs in the row/column indices.

However, to make sure we are all on the same page, perhaps we should briefly discuss this on a call.

**Comment by mhoemmen** [comment]
@egboman wrote:
> For such large systems with large memory then one could simply let local ordinals, global ordinals and offset_type all be int64_t (long long)? This works now, I believe.

There's a submitted pull request to make this work.  I haven't worked on it because 
  1. there is one of me, 
  2. it is a giant blob of commits touching every package downstream of Tpetra, and
  3. we need to come up with tests that would actually prove it works, before unleashing it on the world.

**Comment by mhoemmen** [comment]
@kddevin wrote:
> If getAllValues exists only for backward compatibility, should we be using something else instead to get the three pointers? Is there a mechanism that would always avoid the copy? 

Yes.  Call Tpetra::CrsMatrix::getLocalMatrix to get the KokkosSparse::CrsMatrix.  This only works if the matrix has been fill-completed once.

**Comment by nmhamster** [comment]
@mhoemmen can help you test on the large POWER systems if it's useful. They have lots of memory and great bandwidth (so run tests fast). Unleash that TyrannaTpetra on the world!

**Comment by mhoemmen** [comment]
> Alright, I will leave complaining about it. This might be easy to support as Mark says. However, I hope this is not one off. Extensions like these require long term maintenance, real costs in propagating through the stack, compile times and above all usability.

It's been like this since Chris Baker worked on Tpetra.  I really don't care; it's up to @maherou and @crtrott.  However, I object to your assertion that this is a "one off" [sic] or an "extension."  Mike asked us to do the thing, so we did the thing.  Don't blame me for something that wasn't my idea.

**Comment by mhoemmen** [comment]
The reason the offset type currently depends on `Kokkos::View<..., device_type>::size_type`, is the same reason why `std::vector<..>::size_type == size_t`; it matches the type of the View's dimension.  This relates to the main reason why I hesitated to approve the `LO = long long` pull request: It would not work with `Kokkos::Cuda` right now, because the size_type of Cuda Views is currently `int`.  I could always forbid enabling `LO = long long` in CUDA builds for now, though.

**Comment by mhoemmen** [comment]
Another option would be to use the signed version of View's `size_type`: `std::make_signed<size_type>::type`.  It actually does not matter quite so much for sparse mat-vec performance, though, because all loops in the code use `local_ordinal_type`.  (The loop over the number of entries in _each_ row can use `local_ordinal_type`, because we assume that a single row does not have enough duplicate entries to overflow that type.)

**Comment by mhoemmen** [comment]
@nmhamster wrote:
> @mhoemmen can help you test on the large POWER systems if it's useful. They have lots of memory and great bandwidth (so run tests fast). Unleash that TyrannaTpetra on the world!

*grins* I'm keen on trying it, especially for big graph problems!

**Comment by srajama1** [comment]
@mhoemmen : Nope, not blaming you. I was complaining for the request for this feature :)

**Comment by mhoemmen** [comment]
> @mhoemmen : Nope, not blaming you.

I appreciate that :-) This is a stressful time for us Trilinos folks...

**Comment by mhoemmen** [comment]
I've split off the Tpetra feature request part of this issue into a separate issue, #682.

**Comment by mhoemmen** [comment]
I'm closing this; see #682 for the typedef request.
